Src/VimCore/ModeLineInterpreter.fs

```diff
@@ -0,0 +1,186 @@
+                elif option.StartsWith("no") then
+
+                    // The option toggles the setting off.
+                    let settingName = option.Substring(2)
+                    let toggleOff = (SettingValue.Toggle false)
+                    let setter = fun () ->_localSettings.TrySetValue settingName toggleOff
+                    settingName, setter
+                else
+
+                    // The option toggles the setting on.
+                    let settingName = option
+                    let toggleOn = (SettingValue.Toggle true)
+                    let setter = fun () -> _localSettings.TrySetValue settingName toggleOn
+                    settingName, setter
+
+            // Check whether we should apply the setter.
+            if shouldIgnoreSetting settingName then
+                true
+            elif shouldAllowSetting settingName then
+                setter()
+            else
+                false
+
+        // Split the options string into fields.
+        let splitFields (options: string) =
+            options.Replace(@"\:", ":").Split(' ', '\t')
+
+        // Process the "first" format of modeline, i.e. "vim: ...".
+        let processFirstForm modeLine =
+            let m = Regex.Match(modeLine, _firstPattern)
+            if m.Success then
+                let firstBadOption =
+                    Regex.Matches(m.Groups.[1].Value, _nextGroup)
+                    |> Seq.cast<Match>
+                    |> Seq.map (fun m -> splitFields m.Groups.[1].Value)
+                    |> Seq.concat
+                    |> Seq.tryFind (fun option -> not (processOption option))
+                Some modeLine, firstBadOption
+            else
+                None, None
+
+        // Process the "second" format of modeline, i.e. "vim: set ... :".
+        let processSecondForm modeLine =
+            let m = Regex.Match(modeLine, _secondPattern)
+            if m.Success then
+                let firstBadOption =
+                    splitFields m.Groups.[1].Value
+                    |> Seq.tryFind (fun option -> not (processOption option))
+                Some modeLine, firstBadOption
+            else
+                None, None
+
+        // Try to process either of the two modeline formats.
+        let tryProcessModeLine modeLine =
+            let result = processSecondForm modeLine
+            match result with
+            | Some _, _ -> result
+            | None, _ ->
+               let result = processFirstForm modeLine
+               result
+
+        // Try to process the first few and last few lines as modelines.
+        let tryProcessModeLines modeLines =
+            let lineCount = _textBuffer.CurrentSnapshot.LineCount
+            let snapshot = _textBuffer.CurrentSnapshot
+            seq {
+                yield seq { 0 .. min (modeLines - 1) (lineCount - 1) }
+                yield seq { max modeLines (lineCount - modeLines) .. lineCount - 1 }
+            }

```

Comment:
Isn't this seq just equivalent to all the line numbers in the snapshot?

Reply:
The default for `modelines` is `5`, so typically we search the first five and the last five lines. The min / max calculations here handle files that are shorter than 10 lines long.  Let's say there are 100 lines. Then the expression is like:

```
seq {
    yield seq { 0 .. 4 }
    yield seq { 95 .. 99 }
}
```

Reply:
Added a test to prove that a modeline in the middle of a large file is not processed.